composite_handler/app/api/views.py

```diff
@@ -0,0 +1,133 @@
+                for member in Member.objects.filter(collection=collection)
+            ],
+        }
+
+    def _build_member_response_body(self, member):
+        response = {
+            "id": f"{self._scheme}://{self._hostname}/collections/{member.collection}/members/{member.id}",
+            "status": member.status,
+            "created": member.created_date,
+            "last_updated": member.last_updated_date,
+        }
+
+        if member.image_count:
+            response["image_count"] = member.image_count
+        if member.dlcs_uri:
+            response["dlcs_uri"] = member.dlcs_uri
+        if member.error:
+            response["error"] = member.error
+
+        return response
+
+
+class QueryCollectionAPIView(AbstractAPIView):
+    def get(self, request, *args, **kwargs):
+
+        try:
+            collection = Collection.objects.get(id=kwargs["collection_id"])
+        except Collection.DoesNotExist:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+        self._validate_credentials(collection.customer, request.headers)
+
+        return Response(
+            self._build_collection_response_body(collection),
+            status=status.HTTP_200_OK,
+        )
+
+
+class QueryMemberAPIView(AbstractAPIView):
+    def get(self, request, *args, **kwargs):
+        try:
+            member = Member.objects.get(
+                id=kwargs["member_id"], collection=kwargs["collection_id"]
+            )
+        except Member.DoesNotExist:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+        self._validate_credentials(member.collection.customer, request.headers)
+
+        response_body = self._build_member_response_body(member)
+        response_status = (
+            status.HTTP_301_MOVED_PERMANENTLY
+            if member.status == "COMPLETED"
+            else status.HTTP_422_UNPROCESSABLE_ENTITY
+            if member.status == "ERROR"
+            else status.HTTP_200_OK
+        )
+        response_headers = (
+            {"Location": member.dlcs_uri} if member.status == "COMPLETED" else {}
+        )
+
+        return Response(response_body, response_status, headers=response_headers)
+
+
+class CollectionAPIView(AbstractAPIView):
+    def post(self, request, *args, **kwargs):
+
+        self._validate_credentials(kwargs["customer"], request.headers)
+
+        serializer = CollectionSerializer(
+            data={"json_data": request.data, "customer": kwargs["customer"]}
+        )
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+        collection = serializer.save()
+
+        serializers = [
+            MemberSerializer(data={"json_data": member, "collection": collection.id})
+            for member in collection.json_data["member"]
+        ]
+
+        if not all(serializer.is_valid() for serializer in serializers):
+            return Response(status=status.HTTP_400_BAD_REQUEST)
```

Comment:
Would this 400 response contain what the validation errors are? The collection validation returns `serializer.errors` with response - should this?

Reply:
Ideally yes, however I spent some time looking into an elegant way of doing this and couldn't really find one. I'll take another pass at it.

Reply:
In these views a lot of the behaviour of DRF base classes is being replicated - if they were refactored to use these base view classes, serializers and an authentication class then this would be handled without you having to implement it. I can give you a walkthrough of the approach you could take if you want.

Reply:
@donaldgray taking a second look at this, if there's an error with a member of the collection then that error is returned on line 111. The way the JSON schema has been configured is that this call verifies both the outer collection and the inner members. The subsequent call to every member on line 119 is required as we can't call `.save()` without first calling `.is_valid()`.

So whilst this can technically return a `400`, it never will as any issue with the serialisation of the member would be caught at line 110.